Updated 'social_web_enabled' setting calculation

sagzy opened the pull request:
ref https://linear.app/ghost/issue/PROD-2172

Social web feature is available for users that match all of these criteria:
1) their social web setting is toggled on in Ghost Admin (on by default in Ghost 6.0)
2) their social web labs' flag is toggled on in Ghost Admin (opt-in beta in Ghost 5.0, on-by-default GA in Ghost 6.0) 
3) their site is not hosted on localhost, an IP address, or a subdirectory
4) (Ghost (Pro) only) their access to social web is not limited by config 

Implementation notes:
- given 4), the boot order of Limit Service has changed: it's now initiated before the Settings Service, so that calculated settings such as `social_web_enabled` can use it
- in `@tryghost/limit-service@1.4.0`, we have introduced a new sync method `isDisabled()` to read whether a flag is disabled by config or not. This can be used in replacement of the existing async method `checkWouldGoOverLimit()`, which would require making an async call (and updating CalculatedField / Cache Manager / Settings Service to support async calls!)

sagzy committed:
Updated boot order to init limit service before settings

The calculated setting `socialWebEnabled` will read from the limit service, therefore limit service needs tobe initiated before settings on boot

sagzy committed:
Updated 'social_web_enabled' setting calculation

The calculated setting 'social_web_enabled' is true if:
- the social web setting UI is toggled on
- the social web labs flag is toggled on
- social web is enabled by condig for a given site
- social web is not being accessed from localhost or IP in production

sagzy committed:
Listen to setting UI change to init ActivityPub

sagzy committed:
Refactored for readability

sagzy committed:
Reverted unnecessary logging changed

sagzy commented on ghost/core/core/server/services/settings-helpers/SettingsHelpers.js:
```diff
@@ -220,16 +222,46 @@ class SettingsHelpers {
+        if (this.limitService.limits.limitSocialWeb?.disabled) {
+            logging.warn('Social web is not available for Ghost (Pro) sites without a custom domain, or hosted on a subdirectory');
+            return false;
+        }
```
Ah, if we can safely make that assumption, this simplifies things! @aileen tagging you here as well for visibility, I'll make the change in the meantime

aileen replied:
The short answer is: for a `Flag` limit, yes, this would work. 

However, this would make it impossible to set it as disabled, as we wouldn't be able to check the actual value. The sheer presence of the limit implies that the limit should be enforced. My only concern is that within Ghost, it's not clear that we're dealing with a `Flag` limit here, where this method would work, and we're introducing a pattern that might be used for other limits where it shouldn't 🤔

sagzy replied:
Another solution can be:
- introduce a `isDisabled()` sync method on the Limit Service
- method would return a boolean, reading the `disabled` field, if used for a limit that support it (e.g. Flag)
- method would throw a `IncorrectUsageError`, if used for a limit that doesn't support it (e.g. Max)

sagzy replied:
What do you think of that?

aileen replied:
That would make sense, especially since we're going to have two more limits of that kind in the codebase soon. 

The `errorIfWouldGoOverLimit` is not very useful for a `Flag` limit, and it's strange that it's `async` as it shouldn't need to be. We need to be cautious for the `email` limit, which is a hybrid and _can_ be a numeric or a flag limit, and had to revert those changes. I stumbled over that limit and had failing regression tests, forcing me to revert my changes, when I tried to introduce a change for the `Flag` limit behaviour.

sagzy replied:
Nice that failing tests caught it!

Ok I'll go ahead and open a PR on the limit service with that change then and will request a review :)

sagzy committed:
Used .isLimited() in the limit check

sagzy committed:
Used assert over should

sagzy committed:
Used .isDisabled() method, introduced in @tryghost/limit-service@1.4.0

- .isDisabled() is a sync check available on FlagLimit, that returns undefined if the config is not set, false if the feature is not disabled and true is the feature is disabled

aileen commented on ghost/core/core/server/services/settings-helpers/SettingsHelpers.js:
```diff
@@ -220,16 +222,46 @@ class SettingsHelpers {
+
+        // Social web (ActivityPub) currently does not support Ghost sites hosted on a subdirectory, e.g. https://example.com/blog/
+        const subdirectory = this.urlUtils.getSubdir();
+        if (subdirectory) {
```
We currently have two sites with a subdirectory that have ActivityPub enabled in labs and are on the grandfathered list for that reason. Those two sites won't have the limit applied because they are on that grandfathered list. 

I wonder if we 
- Have to disable the setting for those sites
- Remove them from the grandfather list
- Communicate the reason why we disabled it with them

Probably something to discuss in Slack.

sagzy replied:
Discussed in Slack, resolving

aileen reviewed (approved):
I have added a question, but otherwise LGTM and very readable and maintainable.

ErisDS commented on ghost/core/core/server/services/settings-helpers/SettingsHelpers.js:
```diff
@@ -220,16 +222,46 @@ class SettingsHelpers {
+            return false;
+        }
+
+        return true;
```
I really like the structure of the logic here 👍 

However, I feel like having WARN level logging every time the setting is evaluated is overkill. Using debug would be great here instead and we should trust the UI to communicate with the user.

sagzy replied:
Updated in 9b8297a4901b68c301625995e5fe7f7c7a6afffb

ErisDS reviewed (commented):
The logic all LGTM but I feel the warning logs are unnecessary.

sagzy committed:
Used @tryghost/debug over logging.warn
